core/rust/qdbr/src/parquet_write/symbol.rs

```diff
@@ -1,58 +1,173 @@
-use std::collections::HashMap;
-use std::mem::{self, size_of};
-
+use super::util::BinaryMaxMin;
+use crate::parquet::error::{
+    fmt_err, ParquetError, ParquetErrorCause, ParquetErrorExt, ParquetResult,
+};
+use crate::parquet_write::file::WriteOptions;
+use crate::parquet_write::util;
+use crate::parquet_write::util::{build_plain_page, encode_bool_iter, ExactSizedIter};
 use parquet2::encoding::hybrid_rle::encode_u32;
 use parquet2::encoding::Encoding;
 use parquet2::page::{DictPage, Page};
 use parquet2::schema::types::PrimitiveType;
 use parquet2::write::DynIter;
+use std::char::DecodeUtf16Error;
+use std::cmp::max;
+use std::collections::HashSet;
 
-use crate::parquet_write::file::WriteOptions;
-use crate::parquet_write::util::{build_plain_page, encode_bool_iter, ExactSizedIter};
-use crate::parquet_write::{util, ParquetResult};
+/// Encode the QuestDB symbols to Parquet.
+///
+/// The resulting tuple consists of:
+///   * The parquet dictionary buffer, which is a sequence of the 4-byte-len-prefixed utf8 strings.
+///   * The local keys, which are the indexes into the dictionary buffer.
+///   * The largest key value used, or 0 if no keys were used.
+///
+/// The first argument returned (parquet dict buffer) is encoded in a specific way to be compatible
+/// with QuestDB with zero-read overhead during queries. See `encode_symbols_dict` for details.
+fn encode_symbols_dict(
+    column_vals: &[i32], // The QuestDB symbol column indices (i.e. numeric values).
+    offsets: &[u64],     // Memory-mapped offsets into the QuestDB global symbol table.
+    chars: &[u8], // Memory-mapped global symbol table. Sequence of 4-code-unit-len-prefixed utf16 strings.
+    stats: &mut BinaryMaxMin,
+) -> ParquetResult<(Vec<u8>, Vec<u32>, u32)> {
+    let mut local_keys = Vec::with_capacity(column_vals.len());
```

Comment:
no need to allocate a Vec for local keys.
why not just pass an iterator?

Reply:
Done :-)

It turns out we're still reading the data column thrice.
Refactoring needed to optimise this to a single pass.